### PR TITLE
Ner

### PR DESCRIPTION
**Pull request by mhw32**
- Added Window Featurizer
- Added ConvNet
- Added Name-Entity Tagging using ConvNet
- Added Ada/Core
  - Added AdaOperators for each
  - Added AdaPipelines for 1) word, 2) sentence, 3) text
- Added NLTK parser + lemmatizer (maybe replace in the future_)

### CODE REVIEW

**Comment by mhw32** [comment]
- i did not get a chance to add dropout yet to the MLP layer in CNN

File: common/activation.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+''' Number of activation functions '''
+import autograd.numpy as np
+
+identity = lambda x: x
+
+tanh = lambda x: np.tanh(x)
+
+sigmoid = lambda x: 0.5*(np.tanh(x) + 1.0)
+
+rbf = lambda x: np.exp(-x**2)
```
Comment (TheGoldenRatio) [inline common/activation.py:10]: what does rbf stand for and do?

**Comment by mhw32** [comment]
BTW 95% accuracy on NER (state of the art pretty much!)

File: lemmatizer/lemmatizer_2.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                                  num_epochs=num_epochs,
+                                  step_size=step_size,
+                                  l1_lambda=l1_lambda,
+                                  l2_lambda=l2_lambda,
+                                  nonlinearity=activation.identity)
+
+    return pred_fun, loglike_fun, trained_weights
+
+
+def train_fat_lemmatizer(
```
Comment (TheGoldenRatio) [inline lemmatizer/lemmatizer_2.py:98]: what's fat and thin lemmatizer?

File: lemmatizer/lemmatizer_2.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        all_words.append(word)
+        word_vec = vectorizer(word)
+        lemma_vec = vectorizer(lemma)
+
+        X[i, :] = word_vec
+        y[i, :] = lemma_vec
+
+    word_pos = [p[1]for p in pos_tag(all_words)]
+    pos_set = list(set(word_pos))
+    pos_encoder = one_hot_encoding(pos_set)
```
Comment (TheGoldenRatio) [inline lemmatizer/lemmatizer_2.py:49]: so what are the encodings?

File: common/activation.py (hunk 1 above)
Comment (mhw32) [reply common/activation.py:10]: RBF stands for a radial basis function. How neural networks generally works is that they try to approximate very complicated functions using a number of sequentially stacked non-linear functions. For example, a sigmoid function is a popular one (https://qph.ec.quoracdn.net/main-qimg-05edc1873d0103e36064862a45566dba?convert_to_webp=true): it maps a continuous value to a value between 0 and 1. RBF is just another possible non-linear function. Often it's related to a Gaussian distribution but you should think of it as a way to transform data into a "more interesting" space.

File: lemmatizer/lemmatizer_2.py (hunk 3 above)
Comment (mhw32) [reply lemmatizer/lemmatizer_2.py:49]: In ML, you want to predict a relationship between inputs and outputs. These outputs are often discrete, so for example, outputs could a single digit between 1 and 10, like 1 or 4. You can imagine the neural networks trying to pick a "class" between 1 and 10 for every input. This is often used in image recognition stuff or handwriting recognition. So... when training the neural network, we usually don't just feed the network 4 or 10, we instead "encode" them in an array. 

What a one-hot-encoding is for example (1-indexing of arrays):
1 --> [1, 0, 0, 0, 0, 0, 0, 0, 0, 0]
4 --> [0, 0, 0, 1, 0, 0, 0, 0, 0, 0]
8 --> [0, 0, 0, 0, 0, 0, 0, 1, 0, 0]

The point it to better represent the space for the neural net so it better understands the shape of the outputs.

File: lemmatizer/lemmatizer_2.py (hunk 2 above)
Comment (mhw32) [reply lemmatizer/lemmatizer_2.py:98]: Hmm so a Fat Lemmatizer is just what I call a fully connected neural network where all the nodes have connections to all the other nodes in the next layer. a Thin Lemmatizer is a neural network where node_i in layer_l is only connected to node_i in layer_l+1.

File: core/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                         'NERClassification','DependencyParsing']  # CoRefClassication
+
+        print('[{}] AdaWordPipeline Execution Path:\n{}'.format(
+            str(datetime.now()), '\t\n'.join(step_sequence)))
+
+        embedder_weights = np.load('../storage/word_embedding/glove_weights_300d.npy')
+        with open('../storage/word_embedding/glove_vocab_300d.pkl') as fp:
+            embedder_vocab = cPickle.load(fp)
+
+        pos_tagger_weights = np.load('../storage/pos_tagger/trained_weights.npy')
```
Comment (jesseyli) [inline core/pipeline.py]: switch this to dict(np.load('../storage/pos_tagger/trained_weights.npy'))

File: core/nlp.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+
+class AdaPOSTagger(object):
+    def __init__(self, weights, wordvec):
+        self.weights = weights
+        self.wordvec = wordvec
+
+    def do(self, sentence):
+        ''' sentence : list, list of tokens '''
+        return pos_tagger.predict_from_sentences([sentence])
```
Comment (jesseyli) [inline core/nlp.py]: pos_tagger.predict_from_sentences([sentence], self.weights)

File: models/cnn.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
+''' Conv Nets (similar style to LeNet-5)
```
Comment (jesseyli) [inline models/cnn.py:1]: Are we using this already?

Comment (mhw32) [reply models/cnn.py:1]: No LeNet-5 is just a well known model

**Comment by mhw32** [comment]
Checked by Jonathan and Jesse
